src/raftstore/store/peer_storage.rs

```diff
@@ -743,16 +579,17 @@ fn build_snap_file(f: &mut SnapFile,
     Ok(())
 }
 
-pub fn do_snapshot(snap_dir: &Path,
+pub fn do_snapshot(mgr: SnapManager,
                    snap: &RocksDbSnapshot,
-                   region_id: u64,
-                   ranges: Ranges,
-                   applied_idx: u64,
-                   term: u64)
+                   key: SnapKey,
+                   ranges: Ranges)
                    -> raft::Result<Snapshot> {
-    debug!("begin to generate a snapshot for region {}", region_id);
+    debug!("begin to generate a snapshot for region {}", key.region_id);
 
-    let region: metapb::Region = try!(snap.get_msg(&keys::region_info_key(region_id))
+    let k = key.clone();
+    defer!(mgr.wl().register(k, true));
```

Comment:
seem you will deregister it explicitly after calling this failed, so why not register it directly before calling it?